src/wasm-lib/kcl/src/errors.rs

```diff
@@ -147,6 +147,157 @@ impl KclErrorWithOutputs {
             filenames: Default::default(),
         }
     }
+    pub async fn into_miette_report_with_outputs(
+        self,
+        fs: std::sync::Arc<crate::fs::FileManager>,
+    ) -> anyhow::Result<ReportWithOutputs> {
+        let mut source_ranges = self.error.source_ranges();
+
+        // Pop off the first source range to get the filename.
+        let first_source_range = source_ranges
+            .pop()
+            .ok_or_else(|| anyhow::anyhow!("No source ranges found"))?;
+
+        let filename = self
+            .filenames
+            .get(&first_source_range.module_id())
+            .map(|p| p.to_string())
+            .ok_or_else(|| {
+                anyhow::anyhow!(
+                    "Could not find filename for module id: {:?}",
+                    first_source_range.module_id()
+                )
+            })?;
+        let kcl_source = std::str::from_utf8(&fs.read(&filename, Default::default()).await?)?.to_string();
+
+        let mut related = Vec::new();
+        for source_range in source_ranges {
+            let module_id = source_range.module_id();
+            let filename = self
+                .filenames
+                .get(&module_id)
+                .map(|p| p.to_string())
+                .ok_or_else(|| anyhow::anyhow!("Could not find filename for module id: {:?}", module_id))?;
+            let error = self.error.override_source_ranges(vec![source_range]);
+            let report = Report {
+                error,
+                kcl_source: std::str::from_utf8(&fs.read(&filename, Default::default()).await?)?.to_string(),
+                filename,
+            };
+            related.push(report);
+        }
+
+        Ok(ReportWithOutputs {
+            error: self,
+            kcl_source,
+            filename,
+            related,
+        })
+    }
+}
+
+impl IntoDiagnostic for KclErrorWithOutputs {
+    async fn to_lsp_diagnostics(&self, code: &str) -> Vec<Diagnostic> {
+        let message = self.error.get_message();
+        let source_ranges = self.error.source_ranges();
+
+        let mut diagnostics = Vec::new();
+
+        for source_range in &source_ranges {
```

Comment:
Could map rather than use a for loop, I think

Reply:
i think that would be harder to read no?

Reply:
To me if there is a for loop it, suggests that we're building multiple outputs for a single input, so code which uses a for loop without that is a bit confusing (like I spend time figuring out why it's not a map)

Reply:
ah okay i can convert

Reply:
done!